tools/cndump/kvset_dump.c

```diff
@@ -0,0 +1,385 @@
+        "  -h            print help\n"
+        "  -v            verbose\n"
+        "  <kvdb_home>   KVDB home directory\n"
+        "  <kvset_id>    kvset ID (from the CNDB log)\n");
+
+    if (!global_opts.verbose) {
+        printf("use '-hv' for more detail\n");
+        return;
+    }
+
+    printf("\nDump a KVSET on standard output. Kvset IDs can be seen in the\n"
+        "output of '%s %s'.\n", progname, CNDB_COMMAND_NAME);
+}
+
+static void
+parse_args(int argc, char **argv)
+{
+    const char *arg;
+    int c;
+
+    while ((c = getopt(argc, argv, "+:hv")) != -1) {
+        switch (c) {
+        case 'h':
+            global_opts.help = true;
+            break;
+        case 'v':
+            global_opts.verbose = true;
+            break;
+        case ':':
+            syntax("option -%c requires a parameter", optopt);
+            break;
+        default:
+            syntax("invalid %s option: -%c", KVSET_COMMAND_NAME, optopt);
+            break;
+        }
+    }
+
+    if (global_opts.help) {
+        help();
+        exit(0);
+    }
+
+    if (optind == argc)
+        syntax("missing <kvdb_home> parameter");
+
+    opts.home = argv[optind++];
+
+    if (optind == argc)
+        syntax("missing <kvset_id> parameter");
+
+    arg = argv[optind++];
+    if (parse_u64(arg, &opts.kvset_id))
+        syntax("invalid kvset ID: '%s' (cannot convert to uint64_t)", arg);
+
+    if (optind != argc)
+        syntax("unexpected parameter: '%s'", argv[optind]);
+}
+
+static char *
+fmt_data(void **buf_ptr, size_t *bufsz_ptr, const void *data, size_t data_len, size_t limit)
+{
+    char *buf = *buf_ptr;
+    size_t bufsz = *bufsz_ptr;
+    size_t len = (limit && limit < data_len) ? limit : data_len;
+    size_t output_len = fmt_hex_max_output_len(len);
+
+    if (bufsz < output_len) {
+        void *p = realloc(buf, output_len);
+        if (!p)
+            fatal("realloc", merr(ENOMEM));
+        buf = *buf_ptr = p;
+        bufsz = *bufsz_ptr = output_len;
+    }
+
+    fmt_hex(buf, bufsz, data, len);
+
+    return buf;
+}
+
+static void
+dump_props(const struct mblock_props *p)
+{
+    printf("mbid 0x%lx alen %9u wlen %9u mclass %u\n",
+        p->mpr_objid, p->mpr_alloc_cap, p->mpr_write_len, p->mpr_mclass);
+}
+
+static void
+dump_kvset_props(struct dump_kvset *kvset)
+{
+
+    struct cndb_rec_kvset_add *rec = &kvset->rec.rec.kvset_add;
+
+    printf("\n## mblock properties\n");
+    printf("hblk %2u : ", 0);
+    dump_props(&kvset->hblk.props);
+
+    for (uint i = 0; i < rec->kblkc; i++) {
+        printf("kblk %2u : ", i);
+        dump_props(&kvset->kblkv[i].props);
+    }
+
+    for (uint i = 0; i < rec->vblkc; i++) {
+        printf("vblk %2u : ", i);
+        dump_props(&kvset->vblkv[i].props);
+    }
+}
+
+static void
+dump_kblk(const void *kblock)
+{
+    const struct kblock_hdr_omf *p = kblock;
+    const char *                 base = kblock;
+    const struct wbt_hdr_omf *   wbt_hdr = (void*)p + omf_kbh_wbt_hoff(p);
```

Comment:
This seems to be doing pointer arithmetic on a `void *` which is a GNU extension